notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb

```diff
@@ -0,0 +1,615 @@
+    "                                                                     vocab=vocab)\n",
+    "    # convert entity string to tokens\n",
+    "    entity_tokens, _ = tokens.text_to_tokens(text=entity.lower(), vocab=vocab)\n",
+    "\n",
+    "    results = []\n",
+    "    # iterate through different parts of context\n",
+    "    for network_input, padding, start_idx in prepare_input(entity_tokens=entity_tokens,\n",
+    "                                                           context_tokens=context_tokens):\n",
+    "        # get output layers\n",
+    "        output_start_key = compiled_model.output(\"output_s\")\n",
+    "        output_end_key = compiled_model.output(\"output_e\")\n",
+    "\n",
+    "        # openvino inference\n",
+    "        result = compiled_model(network_input)\n",
+    "        # postprocess the result getting the score and context range for the answer\n",
+    "        score_start_end = postprocess(output_start=result[output_start_key][0],\n",
+    "                                      output_end=result[output_end_key][0],\n",
+    "                                      entity_tokens=entity_tokens,\n",
+    "                                      context_tokens_start_end=context_tokens_start_end,\n",
+    "                                      padding=padding,\n",
+    "                                      start_idx=start_idx)\n",
+    "        results.append(score_start_end)\n",
+    "\n",
+    "    # find the highest score\n",
+    "    answer = max(results, key=operator.itemgetter(0))\n",
+    "    # return the part of the context, which is already an answer\n",
+    "    return context[answer[1]:answer[2]], answer[0]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "### Main Processing Function\n",
+    "\n",
+    "Run entity extraction on specific knowledge base and iterate through the template entities. Final output is a JSON object with two fields i.e. Extraction (consists of Entity, Type and Confidence Score) and Overall Processing time. Currently application supports only few entities from healthcare domain, more templates and entities can be added."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    },
+    "test_replace": {
+     "input()": "\"What does OpenVINO mean?\"",
+     "while True": "for i in range(1)"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "healthcare_template = [\"name\", \"age\", \"medical condition\", \"medication\", \"dosage\", \"dosage unit\"]\n",
+    "\n",
+    "def run_analyze_entities(source, source_type = \"text\"):\n",
+    "    print(f\"Context: {source}\\n\", flush=True)\n",
+    "    \n",
+    "    try:\n",
+    "        context = load_context(source, source_type)\n",
+    "    except:\n",
+    "        print(\"Error: OCR issue in reading file and loading context\")\n",
```

Comment:
This clause is too broad. It's better to catch TesseractNotFoundError. Also the message should be more meaningful like "To extract entities from image you need to install Tesseract-OCR engine". Maybe also some link how to install it?

Reply:
@adrianboguszewski Thanks, makes sense. Is it fine if I modify the message as follows?

        print("OCR Error: To extract entities from image you need to install Tesseract-OCR engine.
              "Follow the installation steps mentioned here "
              "https://stackoverflow.com/questions/50655738/how-do-i-resolve-a-tesseractnotfounderror")

Reply:
As TesseractNotFoundError belong to this class pytesseract.pytesseract.TesseractNotFoundError, so I created a custom exception of the same name to raise at OCR call. And then caught this exception TesseractNotFoundError during entity analysis. Please review if this looks good, thanks!